### PR TITLE
Implement Req/Resp SSZ protocol header and response codes

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Req/Resp messages need to be prefixed with the expected message length.
Responses need to be further prefixed with a status code indicating success or failure.
Also need to delay closing Netty channels until the last write has actually completed.

### CODE REVIEW

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/rpc/encodings/SszEncoding.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    final Bytes header = writeVarInt(payload.size());
+    return Bytes.concatenate(header, payload);
+  }
+
+  @Override
+  public <T> T decodeMessage(final Bytes message, final Class<T> clazz) {
+    try {
+      final CodedInputStream in = CodedInputStream.newInstance(message.toArrayUnsafe());
+      final int expectedLength = in.readRawVarint32();
+      final Bytes payload = Bytes.wrap(in.readRawBytes(expectedLength));
```
Comment (mbaxter) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/rpc/encodings/SszEncoding.java:38]: > If there are any length assertions for length N, it should read exactly N bytes from the stream, at which point an EOF should arise (no more bytes). Should this not be the case, it should be treated as a failure.

Should we throw if the message is too long?

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/rpc/RpcMessageHandler.java [hunk 2]
```diff
@@ -130,20 +131,15 @@ public ResponderHandler(Connection connection) {
     protected void channelRead0(ChannelHandlerContext ctx, ByteBuf byteBuf) {
       STDOUT.log(Level.DEBUG, "Responder received " + byteBuf.array().length + " bytes.");
       Bytes bytes = Bytes.wrapByteBuf(byteBuf);
-      TRequest request = RpcCodec.decode(bytes, method.getRequestType());
+      TRequest request = rpcCodec.decodeRequest(bytes, method.getRequestType());
```
Comment (mbaxter) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/rpc/RpcMessageHandler.java:134]: > If steps (1), (2), or (3) fail due to invalid, malformed, or inconsistent data, the responder MUST respond in error. Clients tracking peer reputation MAY record such failures, as well as unexpected events, e.g. early stream resets.

If decoding fails, should we respond with an error?

**Review by mbaxter — APPROVED** [review]
A couple of questions - otherwise LGTM 🎉

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/rpc/encodings/SszEncoding.java (hunk 1 above)
Comment (ajsutton) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/rpc/encodings/SszEncoding.java:38]: Yes, there's a bunch of error handling and validation to be done at this layer. Logged BC-87.

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/rpc/RpcMessageHandler.java (hunk 2 above)
Comment (ajsutton) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/rpc/RpcMessageHandler.java:134]: Yes.  Logged BC-88.

**Pull request merged**